src/zope/interface/interface.py

```diff
@@ -162,27 +165,120 @@ def implementedBy(self, cls):
+
+    def _compare(self, other):
+        """
+        Compare *self* to *other* based on ``__name__`` and ``__module__``.
+
+        Return 0 if they are equal, return 1 if *self* is
+        greater than *other*, and return -1 if *self* is less than
+        *other*.
+
+        If *other* does not have ``__name__`` or ``__module__``, then
+        return ``NotImplemented``.
+
+        .. caution::
+           This allows comparison to things well outside the type hierarchy,
+           perhaps not symmetrically.
+
+           For example, ``class Foo(object)`` and ``class Foo(Interface)``
+           in the same file would compare equal, depending on the order of
+           operands. Writing code like this by hand would be unusual, but it could
+           happen with dynamic creation of types and interfaces.
+
+        None is treated as a pseudo interface that implies the loosest
+        contact possible, no contract. For that reason, all interfaces
+        sort before None.
+        """
+        if other is self:
+            return 0
+
+        if other is None:
+            return -1
+
+        n1 = (self.__name__, self.__module__)
+        try:
+            n2 = (other.__name__, other.__module__)
+        except AttributeError:
+            return NotImplemented
+
+        # This spelling works under Python3, which doesn't have cmp().
+        return (n1 > n2) - (n1 < n2)
+
+    def __lt__(self, other):
+        c = self._compare(other)
+        if c is NotImplemented:
+            return c
+        return c < 0
+
+    def __le__(self, other):
+        c = self._compare(other)
+        if c is NotImplemented:
+            return c
+        return c <= 0
+
+    def __gt__(self, other):
+        c = self._compare(other)
+        if c is NotImplemented:
+            return c
+        return c > 0
+
+    def __ge__(self, other):
+        c = self._compare(other)
+        if c is NotImplemented:
+            return c
+        return c >= 0
+
+
 @_use_c_impl
-class InterfaceBase(object):
+class InterfaceBase(NameAndModuleComparisonMixin, SpecificationBasePy):
     """Base class that wants to be replaced with a C base :)
     """
 
-    __slots__ = ()
+    __slots__ = (
+        '__name__',
+        '__ibmodule__',
+        '_v_cached_hash',
+    )
+
+    def __init__(self, name=None, module=None):
+        self.__name__ = name
+        self.__ibmodule__ = module
 
     def _call_conform(self, conform):
         raise NotImplementedError
 
+    @property
+    def __module_property__(self):
+        # This is for _InterfaceMetaClass
+        return self.__ibmodule__
+
     def __call__(self, obj, alternate=_marker):
         """Adapt an object to the interface
         """
         try:
             conform = getattr(obj, '__conform__', None)
-        except:
+        # XXX: Do we really want to catch BaseException? Shouldn't
+        # things like MemoryError, KeyboardInterrupt, etc, get through?
+        except: # pylint:disable=bare-except
```

Comment:
yeah, this should be `except Exception:`

Reply:
That'll take some changes in the C code too. There are several cases like that; some [I fixed already](https://github.com/zopefoundation/zope.interface/commit/97997956d1a17a0cfb064dfe0487291e8f8ee181). And when I say fixed, I mean, "made the Python code do `except:` like C  😢 )

Reply:
If I get it right: Technically there is a difference in C between ``except Exception`` and a bare ``except``?
But is there a difference semantically?

Reply:
Compatibility with C code is a good argument for not touching this right now.  (But it should be mentioned in the comment, next to the XXX -- something like "if you fix it here, please also do that in the_right_function() in filename.c".)

Reply:
> Technically there is a difference in C between `except Exception` and a bare `except`?

Yes, much as in Python (where "bare `except:`" is a shortcut for `except BaseException:`). In C, a "bare `except`" is often not even explicitly written. If some function returns `NULL` (the universal indicator of failure), and that `NULL` result is ignored and the error is thrown away with `PyErr_Clear()`, that's basically like writing `except: pass` in Python. That looks like this:
```C
  conform = PyObject_GetAttr(obj, str__conform__);
  if (conform != NULL) {
      return PyObject_CallMethodObjArgs(self, str_call_conform,
                                           conform, NULL);
  }
  else { // except:
    PyErr_Clear(); // pass
  }
  adapter = __adapt__(self, obj);
```

Catching specific exceptions is a bit more difficult and looks something like:
```C
 othername = PyObject_GetAttrString(other, "__name__");
 if (!othername) {
     if (PyErr_Occurred() 
          && PyErr_ExceptionMatches(PyExc_AttributeError)) {
         PyErr_Clear();
     }
 }
```

Which is the rough equivalent of 
```python
try:
    othername = other.__name__
except AttributeError:
    pass
```